File: examples/cvrp_python/columngen/pricer.py
from bin.wrapper import PWSolver

def printKV(key, value):
    print("  " + key.ljust(24) + ": " + str(value))

class Pricer:
    def __init__(self, instance_path):
        self.instance_path = instance_path

        self.pathwyse = PWSolver()
        self.pathwyse.readProblem(instance_path)
        self.pathwyse.setupAlgorithms()
        self.pathwyse.setAlgorithmsParallel(True)
        self.using_heuristics = True
        self.using_exact = False
        self.reset_level = 0
        self.mu = None
        self.gamma = 0.0
        self.useHeuristics()
        self.debug = False

    def isUsingHeuristics(self):
        return self.using_heuristics
    
    def isUsingExact(self):
        return self.using_exact
    
    def useHeuristics(self):
        self.using_heuristics = True
        self.using_exact = False
        self.pathwyse.enableAllAlgorithms()
        self.pathwyse.disableAlgorithm(0)

    def useExact(self):
        self.using_heuristics = False
        self.using_exact = True
        self.pathwyse.disableAllAlgorithms()
        self.pathwyse.enableAlgorithm(0)

    def updatePricers(self, duals):
        mu = [-m for m in duals[0]]
        self.mu = mu
        self.pathwyse.setNodeCosts(mu)

        gamma = -1*duals[1]
        self.gamma = gamma
        self.pathwyse.setInitCost(gamma)
        for algo_id in self.pathwyse.getEnabledAlgorithms():
            self.pathwyse.resetAlgorithm(algo_id, self.reset_level)
        return

    def solve(self):
        self.pathwyse.solve()
        return
        
    # Recomputes the reduce costs of a column to avoid issues (PW obj and the true RC might differ, depending on scaling precision)
    def recomputeRC(self, cost, col):
        rc = cost + sum(self.mu[x] for x in col) + self.gamma
        return rc
        
    def collectColumns(self):
        threshold = 0
        bestRC = 0
        costs = list()
        columns = list()

        nsol = self.pathwyse.getNumberOfSolutions()
        for i in range(nsol):
            pw_rc = self.pathwyse.getSolutionObjective(i)
            cost = self.pathwyse.getSolutionArcCost(i)
            col = self.pathwyse.getSolutionTour(i)
            rc = self.recomputeRC(cost, col)
            diff = abs(rc - pw_rc)
            if self.debug:
                printKV("PathWyse RC", pw_rc)
                printKV("Recomputed RC", rc)
                printKV("OBJ diff", diff)
                printKV("Column cost", cost)
                print("")

            if diff > 1:
                print("Warning: RC and PW objective differ by more than 1")
            
            if rc < threshold:
                costs.append(cost)
                columns.append(col)

            if rc < bestRC:
                bestRC = rc

        return bestRC, costs, columns
    
    def getNumNodes(self):
        return self.pathwyse.getNumberOfNodes()

    def clearColumns(self):
        self.pathwyse.clearSolutions()

File: examples/cvrp_cpp/pricer.cpp
#include "pricer.h"
#include "console.h"
#include <cmath>

Pricer::Pricer(const std::string& instance_path)
    : using_heuristics(true), using_exact(false), reset_level(0), gamma(0.0) {
    pathwyse.readProblem(instance_path);
    pathwyse.setupAlgorithms();
    pathwyse.setAlgorithmsParallel(true);
    useHeuristics();
}

void Pricer::useHeuristics() {
    using_heuristics = true;
    using_exact = false;
    pathwyse.enableAllAlgorithms();
    pathwyse.disableAlgorithm(0);
}

void Pricer::useExact() {
    using_heuristics = false;
    using_exact = true;
    pathwyse.disableAllAlgorithms();
    pathwyse.enableAlgorithm(0);
}

void Pricer::updatePricers(const Duals& duals) {
    mu.assign(duals.mu.size(), 0.0);
    for (size_t i = 0; i < duals.mu.size(); ++i)
        mu[i] = -duals.mu[i];
    pathwyse.setNodeCosts(mu);

    gamma = -duals.gamma;
    pathwyse.setInitCost(gamma);

    for (int algo_id : pathwyse.getEnabledAlgorithms())
        pathwyse.resetAlgorithm(algo_id, reset_level);
}

void Pricer::solve() {
    pathwyse.solve();
}

// Recomputes the reduced cost of a column to avoid issues (PW obj and the
// true RC might differ, depending on scaling precision)
double Pricer::recomputeRC(double cost, const std::vector<int>& col) const {
    double rc = cost + gamma;
    for (int x : col)
        rc += mu[x];
    return rc;
}

Pricer::CollectResult Pricer::collectColumns() {
    CollectResult result;
    result.best_rc = 0.0;
    const double threshold = 0.0;

    int nsol = pathwyse.getNumberOfSolutions();
    for (int i = 0; i < nsol; ++i) {
        double pw_rc = pathwyse.getSolutionObjective(i);
        double cost = pathwyse.getSolutionArcCost(i);
        std::vector<int> col = pathwyse.getSolutionTour(i);
        double rc = recomputeRC(cost, col);
        double diff = std::abs(rc - pw_rc);

        if (debug) {
            printKV("PathWyse RC", pw_rc);
            printKV("Recomputed RC", rc);
            printKV("OBJ diff", diff);
            printKV("Column cost", cost);
            std::cout << "\n";
        }

        if (diff > 1)
            std::cout << "Warning: RC and PW objective differ by more than 1\n";

        if (rc < threshold) {
            result.costs.push_back(cost);
            result.columns.push_back(col);
        }
        if (rc < result.best_rc)
            result.best_rc = rc;
    }

    return result;
}

int Pricer::getNumNodes() {
    return pathwyse.getNumberOfNodes();
}

void Pricer::clearColumns() {
    pathwyse.clearSolutions();
}
